src/Analysis/Engine/Test/LineFormatterTests.cs

```diff
@@ -0,0 +1,472 @@
+        public void LineContinuation() {
+            AssertSingleLineFormat("a+b+ \\\n", "a + b + \\");
+        }
+
+        [TestMethod, Priority(0)]
+        public void MultilineChainedCall() {
+            const string code = "foo.a() \\\n   .b() \\\n   .c()";
+            AssertSingleLineFormat(code, "foo.a() \\");
+            AssertSingleLineFormat(code, ".b() \\", line: 2, editStart: 4);
+            AssertSingleLineFormat(code, ".c()", line: 3, editStart: 4);
+        }
+
+        [TestMethod, Priority(0)]
+        public void BracketCommas() {
+            AssertSingleLineFormat("a[:, :, :, 1]");
+            AssertSingleLineFormat("a[x:y, x + 1 :y, :, 1]");
+            AssertSingleLineFormat("a[:, 1:3]");
+            AssertSingleLineFormat("a[:, :3, :]");
+            AssertSingleLineFormat("a[:, 3:, :]");
+        }
+
+        [TestMethod, Priority(0)]
+        public void MultilineStringTrailingComment() {
+            AssertSingleLineFormat(@"'''
+foo
+''' # comment", "  # comment", line: 3, editStart: 4);
+        }
+
+        [TestMethod, Priority(0)]
+        public void Backtick() {
+            AssertSingleLineFormat("`a`", languageVersion: PythonLanguageVersion.V27);
+            AssertSingleLineFormat("foo(`a`)", languageVersion: PythonLanguageVersion.V27);
+            AssertSingleLineFormat("`a` if a else 'oops'", languageVersion: PythonLanguageVersion.V27);
+        }
+
+        [TestMethod, Priority(0)]
+        public void ExecStatement() {
+            AssertSingleLineFormat("exec code", languageVersion: PythonLanguageVersion.V27);
+            AssertSingleLineFormat("exec (code)", languageVersion: PythonLanguageVersion.V27);
+            AssertSingleLineFormat("exec(code)");
+        }
+
+        [TestMethod, Priority(0)]
+        public void GrammarFile() {
+            var src = TestData.GetPath("TestData", "Formatting", "pythonGrammar.py");
+
+            string fileContents;
+            using (var reader = new StreamReader(src, true)) {
+                fileContents = reader.ReadToEnd();
+            }
+
+            var lines = fileContents.Split(new[] { "\r\n", "\r", "\n" }, StringSplitOptions.None);
+
+            using (var reader = new StringReader(fileContents)) {
+                var lineFormatter = new LineFormatter(reader, PythonLanguageVersion.V37);
+
+                for (var i = 0; i < lines.Length; i++) {
+                    var lineNum = i + 1;
+
+                    var edits = lineFormatter.FormatLine(lineNum);
+
+                    var lineText = lines[i];
+                    var lineTextOrig = lines[i];
+
+                    foreach (var edit in edits) {
+                        edit.range.start.line.Should().Be(i);
+                        edit.range.end.line.Should().Be(i);
+
+                        lineText = ApplyLineEdit(lineText, edit);
+                    }
+
+                    lineText.Should().Be(lineTextOrig, $"because line {lineNum} should be unchanged");
```

Comment:
`ApplyLineEdit` doesn't refer any product code. What's the purpose of this assert?

Reply:
I'm not sure what you mean. `ApplyLineEdit` applies a `TextEdit` to a single line of text and returns the edited string. I use it to apply the formatter's edit to make sure that it didn't change anything about the text.

Reply:
Ok, is it different from 
```cs
foreach (var edit in edits) {
    var start = edit.range.start;
    var end = edit.range.end;
    start.line.Should().Be(i);
    end.line.Should().Be(i);
    edit.newText.Should().Be(lineText.Substring(start.character, end.character - start.character - 1));
}
```

Reply:
Yes, it's different than that. You couldn't check that each edit works, since they'd technically be applied by the editor in order, and the final result is what should actually be unchanged, but in my case it would probably be more correct to assert that either there are no edits (so therefore no change), or that there's a single edit which doesn't change the line. I could also do something like:

```csharp
var edits = lineFormatter.FormatLine(lineNum);
edits.Should().NotBeNull();

if (edits.Length == 0) {
    continue;
}

edits.Length.Should().Be(1);
var edit = edits[0];

var start = edit.range.start;
var end = edit.range.end;

start.line.Should().Be(i);
end.line.Should().Be(i);

var startIndex = start.character;
var removeCount = end.character - start.character - 1;

var lineText = lines[i];
var newLineText = lineText.Remove(startIndex, removeCount).Insert(startIndex, edit.newText);
newLineText.Should().Be(lineText, $"because line {lineNum} should be unchanged");
```

Which would do the same thing but avoids the helper.

Reply:
Is it to show the whole string in the error message?

Reply:
You're totally right. I hadn't thought of that. I'll commit the changes.

Reply:
Showing the whole string isn't super important, since the only thing that would be cut off is whitespace (which currently does not get replaced, but modifying the formatter to replace it would not break this method of testing).